### PR TITLE
refactor: contract-based redemption

### PR DESCRIPTION
**Pull request by kupermind**
- Contract-based redemption.

### CODE REVIEW

File: contracts/MemeArbitrum.sol [hunk 1]
```diff
@@ -72,9 +72,7 @@ contract MemeArbitrum is MemeFactory {
         FactoryParams memory factoryParams,
         address _l2TokenRelayer,
         address _balancerVault,
-        bytes32 _balancerPoolId,
```
Comment (kupermind) [inline contracts/MemeArbitrum.sol:75]: Cleaning up meme arb

File: contracts/MemeBase.sol [hunk 2]
```diff
@@ -83,6 +97,10 @@ contract MemeBase is MemeFactory {
         address _l2TokenRelayer,
         address _balancerVault,
         bytes32 _balancerPoolId,
+        address _redemptionMemeBase,
+        uint256 _redemptionTokenIdx,
+        string memory redemptionName,
+        string memory redemptionSymbol,
```
Comment (DavidMinarsch) [inline contracts/MemeBase.sol:103]: Can we please just hard code these. There's no reason imo to provide them as args given the whole system is literally defined for this exact token

File: contracts/MemeBase.sol [hunk 3]
```diff
@@ -91,7 +109,29 @@ contract MemeBase is MemeFactory {
...
+            MemeFactory.MemeSummon memory memeSummon = IMemeBase(_redemptionMemeBase).memeSummons(redemptionMemeToken);
+            contributionAGNT = memeSummon.nativeTokenContributed;
+
+            // Redemption amount: collected amount - 10% for burn = 127412857890000000000
+            redemptionAmount = contributionAGNT * 9 / 10;
+
+            // Set up redemption procedure
+            uint8 decimals = IERC20(redemptionMemeToken).decimals();
+            // Half of the supply has been burnt, thus it needs to be increased
+            uint256 totalSupply = 2 * IERC20(redemptionMemeToken).totalSupply();
```
Comment (DavidMinarsch) [inline contracts/MemeBase.sol:130]: This is not safe to do. This was really not useful work. What if someone buys and burns tokens now. Just hardcode it

File: contracts/MemeBase.sol [hunk 4]
```diff
@@ -91,7 +109,29 @@ contract MemeBase is MemeFactory {
...
+            // This is a required check if any accounts burns tokens as well
+            require(totalSupply == 1_000_000_000 ether, "Total supply has changed");
```
Comment (DavidMinarsch) [inline contracts/MemeBase.sol:132]: So we do all the above work to then just have it verified anyway

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**